### PR TITLE
Fix PoW mempool and fees for Atomics

### PR DESCRIPTION
**Pull request by poolcoke**

### CODE REVIEW

File: lib/block_view.go [hunk 1]
```diff
@@ -4420,21 +4420,53 @@ func (bav *UtxoView) ConnectBlock(
...
+				// For PoS, the maximum block reward is based on the maximum utility fee.
+				// Add the utility fees to the max utility fees. If any overflow
+				// occurs mark the block as invalid and return a rule error.
+				maxUtilityFee, err = SafeUint64().Add(maxUtilityFee, utilityFee)
+				if err != nil {
+					return nil, errors.Wrapf(RuleErrorPoSBlockRewardWithInvalidAmount,
+						"ConnectBlock: error computing maxUtilityFee: %v", err)
+				}
+			} else {
+				for _, innerTxn := range txn.TxnMeta.(*AtomicTxnsWrapperMetadata).Txns {
```
Comment (lazynina) [inline lib/block_view.go:4444]: This is quite deeply nested, so cognitive complexity is pretty high in here. Can we write some sort of utility function that given an atomic transaction and a block reward recipient, iterates over it and return the total fees minus any fees paid by the block reward recipient? This function can return both BMF and old PoW fees. This way we can use the same function in each of the three places its used and not have such deeply nested for-loops and conditionals.

**Review by lazynina — CHANGES_REQUESTED** [review]

**Review by lazynina — CHANGES_REQUESTED** [review]
All the logic is correct, so it *works* in my eyes, but it's going to be a little annoying to maintain with the deeply nested logic as we have it. Any modifications you can make to simplify and de-duplicate looping over the inner txns to filter out block reward recipient fees would be very helpful.

**Comment by lazynina** [comment]
* **#1224** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/1224?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1220** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/1220?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `feature/proof-of-stake`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @poolcoke and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: lib/block_producer.go [hunk 2]
```diff
@@ -452,17 +484,34 @@ func RecomputeBlockRewardWithBlockRewardOutputPublicKey(
...
+					glog.Errorf("DeSoMiner._startThread: Error parsing transactor public key: %v", err)
+					continue
+				}
+				if !transactorPublicKey.IsEqual(blockRewardOutputPublicKey) {
+					totalFees, err = SafeUint64().Add(totalFees, innerTxn.TxnFeeNanos)
+					if err != nil {
+						glog.Errorf("DeSoMiner._startThread: Error adding txn fee: %v", err)
+						continue
+					}
+				}
```
[the remark this answers is not in the document]
Comment (poolcoke) [reply lib/block_producer.go:513]: Yeah, I think this is better. We used `!transactorPublicKey.IsEqual(blockRewardOutputPublicKey)` before this change so I think it seemed natural to do but it was confusing while I was reading through it.

File: lib/block_producer.go [hunk 3]
```diff
@@ -305,12 +305,43 @@ func (desoBlockProducer *DeSoBlockProducer) _getBlockTemplate(publicKey []byte)
...
+								"DeSoBlockProducer._getBlockTemplate: problem parsing transactor public key: ")
+					}
+
+					// Block the block reward output public key from getting free transaction fees within
+					// an atomic transaction.
+					if transactorPublicKey.IsEqual(blockRewardOutputPublicKey) {
+						if feeNanos < innerTxn.TxnFeeNanos {
+							return nil, nil, nil,
+								fmt.Errorf("DeSoBlockProducer._getBlockTemplate: " +
+									"feeNanos for atomic transaction underflowed during blow reward check; " +
```
[the remark this answers is not in the document]
Comment (poolcoke) [reply lib/block_producer.go]: Addressed with PR#1224.

File: lib/block_producer.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (poolcoke) [reply lib/block_producer.go]: Addressed with PR#1224.

File: lib/block_view.go (hunk 1 above)
Comment (poolcoke) [reply lib/block_view.go]: Addressed with PR#1224.

**Review by lazynina — APPROVED** [review]

**Review by tholonious — APPROVED** [review]

File: lib/block_view.go [hunk 4]
```diff
@@ -4360,21 +4360,47 @@ func (bav *UtxoView) ConnectBlock(
...
-			// Add the fees from this txn to the total fees. If any overflow occurs
-			// mark the block as invalid and return a rule error. Note that block reward
-			// txns should count as having zero fees.
-			if totalFees > (math.MaxUint64 - currentFees) {
-				return nil, RuleErrorTxnOutputWithInvalidAmount
-			}
-			totalFees += currentFees
+			if txn.TxnMeta.GetTxnType() != TxnTypeAtomicTxnsWrapper {
+				// Compute the BMF given the current fees paid in the block.
+				_, utilityFee = computeBMF(currentFees)
```
Comment (diamondhands0) [inline lib/block_view.go:4365]: Am I crazy or does this call to computeBMF need to be gated behind a block height?

**Review by diamondhands0 — CHANGES_REQUESTED** [review]
See if you need to put _computeBMF behind a block height

**Review by diamondhands0 — APPROVED** [review]

File: lib/block_view.go (hunk 4 above)
Comment (poolcoke) [reply lib/block_view.go:4365]: Gated on line 4449.

**Comment by lazynina** [comment]
> See if you need to put _computeBMF behind a block height

As discussed offline, all downstream uses are confined to pos gated logic.

**Pull request merged**